README.md
# CloudSea

CloudSea 是一种封装了一些常见的业务逻辑的请求库，如请求参数与返回数据处理，业务逻辑码（retcode）处理，失败重试，统一成功失败处理等。默认支持小程序与 WEB 请求，可自定义适配器支持更多不同请求。

## 特征

- baseUrl 设置
- 失败后自动 2 次重启
- 业务返回码 retcode 字段指定及白名单处理
- request 请求参数插件处理
- response 返回数据插件处理
- 完成处理（不论成功或失败），见下面的 completeHandler（可在里面进行 hideLoading 处理、上报处理等）
- 成功或失败的统一 promise 链，见下面的 thenHandler 和 catchHandler
- 支持取消发送请求
- 请求适配器自定义
- Typescript 编写
- 其他定制

## 安装

```bash
npm i cloud-sea --save
```

## 使用

### 光速入门

```ts
import { CloudSea } from 'cloud-sea';

const gRequest = new CloudSea();

const requestData = {
  url: 'xxx',
  data: {
    a: 1,
  },
};

// request 调用，如果没有指定 method，则为 GET 请求
gRequest.request(requestData);

// get 调用
gRequest.get(requestData);

// post 调用
gRequest.post(requestData);
```

### 手把手自定义

**1、实例化**

这里实例化的参数分为两类，一类是用于真正请求的，如 header，timeout 等；另一类是用于辅助控制逻辑的，会统一放到 ext 对象中。

```ts
import { CloudSea, ICloudSeaConfig } from 'cloud-sea';

// 在默认配置参数上扩展新的字段 loadingTips，用于控制是否显示 loading
interface IExt extends ICloudSeaConfig {
  loadingTips: boolean;
}

// ICloudSeaOptions 的结构大概是 {..., ext?: {...}}
// 前面三点省略表示请求的一些字段，后面三点表示 config 字段
// 这里 options.ext 里面的配置会覆盖默认的配置
const gRequest = new CloudSea<IExt>({
  ext: {
    loadingTips: false, // 新增的字段
    logicErrorMsgKey: 'msg', // 覆盖默认的字段
  },
});

// 或通过 setConfig 方法修改默认的配置字段
gRequest.setConfig({
  logicErrorMsgKey: 'msg', // 覆盖默认的字段
});

// 如果没有新增字段，则不需要泛型
// const gRequest = new Request({ logicErrorMsgKey: 'msg' });

export default gRequest;
```

默认的一些用于辅助控制逻辑的参数，如下（可通过 setConfig 方法或实例化自定义参数覆盖）：

```ts
{
  baseUrl: '',
  xRequestId: true, // header 头部加上 x-request-id，方便打通日志链路，默认开启
  repeatNum: 2, // 默认失败自动重试 2 次
  timeout: 0, // 默认请求本身支持 timeout 参数的不需要这个，设置为 0 即可，如果需要兼容不支持的，设置这个即可，不要设置请求本身的，qq 小程序本身不支持 timeout 设置
  retcodeKey: 'retcode', // retcode 字段，false 表示不启用该功能
  retcodeWhiteList: [], // retcode 白名单，默认空数组为retcode为0进入then，其余为进入catch，false 表示不启用该功能，retcodeKey 为 false 也没有该功能
  logicErrorMsgKey: 'message', // ，默认逻辑错误文本字段，支持一层对象，如 errData.msg
  loginErrorMsgUnknown: 'Unknown Error', // 逻辑错误文本默认，如果后台没有返回具体的错误，则显示该文本
  adapter: getDefaultAdapter(), // 自动适配 WEB 还是小程序发送请求
}
```

**2、插件处理入参与返回**

```ts
// 插件式处理请求参数或返回数据
// ------------------------------------------------
// 处理请求参数
// 可使用多个 use，每个 use 函数按注册顺序依次进入管道处理，所有的入参都挂到 ctx.req 对象上，最后返回 ctx
// 如小程序可再次设置 header 的 cookie 字段，用于鉴权
gRequest.req.use((ctx) => {
  // 处理 req ...
  // ctx.req.xxx = xxx;

  // 最后记得 return ctx
  return ctx;
});

// 处理返回数据，所有的返回都挂到 ctx.res 对象上，同上请求参数的处理
gRequest.res.use((ctx) => {
  // 处理 res ...
  // ctx.res = {};

  // 最后记得 return ctx
  return ctx;
});
```

**3、成功失败的统一处理**

```ts
// 处理请求的成功及失败
// ------------------------------------------------
// 请求完成处理，不论成功或失败，可用于关闭 loading，上报等
// err 有三种 type，分别为：逻辑错误，服务器错误，网络错误，具体见错误说明部分
gRequest.completeHandler = (ctx, err?) => {
  // 成功
  if (!err) {
  }

  // 可以根据 err.type 来判断错误类型
  // ------------------------------------
  // fail
  if (err.type === REQUEST_ERROR_MAP.fail) {
  }

  // server
  if (err.type === REQUEST_ERROR_MAP.server) {
  }

  // logic
  if (err.type === REQUEST_ERROR_MAP.server) {
  }
};

// 统一成功处理
gRequest.thenHandler = (ctx) => {
  // 返回下一步进入成功的数据
  return ctx.res;
};

// 统一错误处理：如 toast 提示，上报错误等
gRequest.catchHandler = (err, ctx) => {
  // 处理逻辑
};
```

**4、函数封装**

```ts
// 常用 get 与 post 方法的进一步封装
// ------------------------------------------------
// 先定义返回的数据格式
export interface IRequestRes<T> {
  retcode: number;
  data: T;
  cost: number;
  message: string;
}
// 简化 get 方法
// 如果实例化有泛型，则这里的 IRequestOptions 也需要 <IExt>，否则不需要
export function gGet<T>(data: IRequestOptions<IExt>) {
  return gRequest.get<IRequestRes<T>>(data);
}

// 简化 post 方法
// 如果实例化有泛型，则这里的 IRequestOptions 也需要 <IExt>，否则不需要
export function gPost<T>(data: IRequestOptions<IExt>) {
  return gRequest.post<IRequestRes<T>>(data);
}
```

**5、具体发送请求**

导入 `gRequest.ts`，调用其方法发送具体请求

```ts
// api.ts
// ------------------------------------
import { IRequestOptions, ICloudSeaConfig } from 'cloud-sea';
import gRequest, { gGet, gPost } from './gRequest';

// 定义返回结构
interface Res<T> {
  retcode: number;
  data: T;
  message?: string;
}

// 常用的请求参数类型如下，具体的话可见类型提示：
// {
//   url: string;
//   method?: IMethod;
//   header?: Record<string, string>;
//   data?: IAnyObject | ...; // 这个有多种值，xhr 这里只做做了 IAnyObject 的特殊处理，其余全部透传
//   timeout?: number; // 注意该 timeout 为默认支持的，如果确认你要兼容的都支持，那么就可以考虑去掉 ext 中的 timeout
//   ext?: {...}; // 上面说的配置
//   task?: { name: string} // 用于取消请求用
// }

//

// 请求 1
interface Data1 {
  isValid: boolean;
}

// 使用 request 方法发送
gRequest.request<Res<Data1>>({
  url: 'xxx',
  method: 'POST',
  header: {
    // 设置 header
    'x-language': 'en', // 设置请求语言
  },
  data: {
    a: 1,
    b: 2,
  },
  // 可覆盖前面的设置
  ext: {
    repeatNum: 0, // 失败不重试
  },
});

// 或直接采用 gPost 方法发送
gPost<Res<Data1>>({
  url: 'xxx',
  header: {
    // 设置 header
    'x-language': 'en', // 设置请求语言
  },
  data: {
    a: 1,
    b: 2,
  },
  // 可覆盖前面的设置
  ext: {
    repeatNum: 0, // 失败不重试
  },
});

// get 请求
interface Data2 {
  name: string;
}
gGet<Res<Data2>>({
  url: 'xxx',
  // 自动拼成 query
  data: {
    a: 1,
    b: 2,
  },
  ext: {
    retcodeKey: 'code', // 这条请求的 retcodeKey 是 code 字段
  },
});

// post 请求
interface Data3 {
  name: string;
  id: string;
}
gPost<Res<Data3>>({
  url: 'xxx',
  data: {
    a: 1,
    b: 2,
  },
  ext: {
    retcodeWhiteList: [3455, 6784], // retcode 为 0， 3455，6784 将会按成功处理，其余按失败处理
  },
});
```

**6、取消正在发送中的请求**

```ts
// 给请求取个名字，注意只有请求有了 name 值才能被取消，否则不能被取消
interface Data3 {
  name: string;
  id: string;
}
gPost<Res<Data3>>({
  url: 'xxx',
  task: {
    name: 'hello',
  },
});

// 取消单个 taskName 为 hello 的请求
gRequest.abort('hello');

// 取消所有有请求名的正在发送中的请求
gRequest.abort();
```

## ctx 参数说明

ctx 由 `req、res、ext` 三大属性组成， 其 TS 类型如下：

PS：注意所有的用于辅助功能的参数都挂到 ext 上，不要随便在 req 上面挂属性。req 会透传到请求适配器，用于发送请求的所有参数。

```ts
export interface ICloudSeaConfig {
  baseUrl: string; // 基础 url，以 https 开头
  repeatNum: number; // 请求失败重试次数
  xRequestId: boolean; // 是否生成请求 id
  timeout: number; // 超时时间，如果确定发请求的 api 本身就支持 timeout 属性，可以设置该值为 0
  retcodeKey: false | string | string[]; // retcode 字段，false 表示不启用该功能
  retcodeWhiteList: false | number[]; // retcode 白名单，默认 0 和 白名单表示业务成功，其余为失败，false 表示不启用该功能。
  logicErrorMsgKey: string; // 业务逻辑错误文本字段
  logicErrorMsgUnknown: string; // 默认的业务逻辑错误文本，如果后台没有返回对应的错误信息，则将使用该信息
  adapter: IAdapter | undefined; // 请求 adapter
}

export interface IRequestTask {
  name?: string; // 用于取消请求
  client: any;
  urlHasNoSearch: string; // 去掉请求 url 的 query，可用于上报请求地址
  timer: ReturnType<typeof setTimeout>;
  costTime: number; // 请求总共花费时间，当 xRequestTime 为 true，则有该值
  repeatTry: () => Promise<IAnyObject>; // 用于重试
}

export interface IRequestCtx<T extends ICloudSeaConfig> {
  req: IReqOptions & { header: Record<string, string> };
  res: IRequestSuccessCallbackResult | Record<string, never>;
  ext: T;
  // ext: U extends Record<string, never> ? ICloudSeaConfig : ICloudSeaConfig & U;
  task: IRequestTask;
}

export type IRequestOptions<T extends Partial<ICloudSeaConfig>> = IReqOptions & {
  ext?: T;
} & { task?: { name: string } };

export type ICloudSeaOptions<T extends Partial<ICloudSeaConfig>> = Omit<IRequestOptions<T>, 'url' | 'data' | 'task'>;
```

## 请求错误说明

```ts
import { REQUEST_ERROR_MAP, RequestError, DEFAULT_LOGIC_ERROR_MSG_UNKNOWN } from 'cloud-sea';

// REQUEST_ERROR_MAP
// {
//   fail: 'REQUEST_ERROR_FAIL', // 直接 fail 的错误
//   server: 'REQUEST_ERROR_SERVER', // 服务端错误，statusCode 小于 200，大于等于 300
//   logic: 'REQUEST_ERROR_LOGIC', // 业务逻辑错误
// }
```

1、第一种错误：根本没有收到服务端返回的信息，这就有了 fail 的错误（以小程序来说，就是 fail 的回调，以 web 来说，就是 XMLHttpRequest 的 onabort, ontimeout, onerror 事件触发），该错误抛出：`new RequestError(err.message || err.errMsg, { type: REQUEST_ERROR_MAP.fail })`。

2、第二种错误：接受到了服务端的信息，但是 statusCode 小于 200，大于等于 300，这就有了 server 错误，该错误抛出： `new RequestError('Request Server Error', { type: REQUEST_ERROR_MAP.server, statusCode });`。

3、第三种错误：虽然 statusCode 大于等于 200，小于 300，但是如果用户没有登录，或身份不对等都无法获取到正确的数据，所以就有了 logic 错误，该错误抛出： `new RequestError(logicErrMsg, { type: REQUEST_ERROR_MAP.logic, retcode });`。`logicErrMsg` 的取值见下面的说明。

这三种抛出的错误默认都会进入到 `catchHandler` 进行处理，如果对 `retcodeWhiteList` 进行设置，则第三种的 logic 错误白名单内的会当做成功进入到 `thenHandler` 处理。

`catchHandler` 和 `completeHandler` 的 err 参数，就是上面的三种错误实例。

### `logicErrMsg`

`logicErrMsg` 是通过调用 `getLogicErrMsg` 方法得到的，逻辑如下：

```ts
getLogicErrMsg(ctx: IRequestCtx): string {
  const { res, ext } = ctx;
  const { logicErrorMsgKey, logicErrorMsgUnknown } = ext;

  // 如果没有指定 msg 的 key
  if (!logicErrorMsgKey) {
    return logicErrorMsgUnknown;
  }

  // 如果有指定，则取该值
  let logicErrMsg: string | undefined = res.data?.[logicErrorMsgKey];
  // 错误信息可能不是一个直接的 string 字段，而是被包裹在一个对象中，如 errData: { text: 'xxx', code: xxx };
  // 这样可以设置 key 值为 'errData.text'，分割得到最终的字段
  if (!logicErrMsg && logicErrorMsgKey.includes('.')) {
    const [key1, key2] = logicErrorMsgKey.split('.');
    if (res.data?.[key1]?.[key2]) {
      logicErrMsg = res.data[key1][key2];
    }
  }

  return logicErrMsg || logicErrorMsgUnknown;
}
```

## 实战

```ts
import { CloudSea, RequestError, ICloudSeaConfig, IRequestOptions } from 'cloud-sea';
// 如为小程序，可直接使用 `wx.showLoading` 与 `wx.showToast` 来实现。
import { showLoading, hideLoading, showToast } from 'x-global-api';
import humps from 'humps';

interface IExt extends ICloudSeaConfig {
  loadingTips: boolean; // 请求发送时 showLoading
  camelCase: boolean; // 失败 showToast
  failToast: boolean; // 转驼峰处理
}

const gRequest = new CloudSea<IExt>({
  ext: {
    loadingTips: false,
    camelCase: true,
    failToast: true,
  },
});

gRequest.req.use((ctx) => {
  // 默认如果 loadingTips 为 true，则显示 loading
  if (ctx.ext.loadingTips) {
    showLoading({ title: '数据加载中...', mask: true });
  }

  return ctx;
});

gRequest.res.use((ctx) => {
  // 默认如果 camelCase 为 true，则自动进行转驼峰处理
  if (ctx.ext.camelCase) {
    ctx.res.data = humps.camelizeKeys(ctx.res.data);
  }

  return ctx;
});

// 不管成功，失败都要 hideLoading
gRequest.completeHandler = (ctx, err?) => {
  if (ctx.ext.loadingTips) {
    hideLoading();
  }
};

// 统一成功处理
gRequest.thenHandler = (ctx) => {
  // 返回下一步进入成功的数据
  return ctx.res.data;
};

// 失败错误提示
gRequest.catchHandler = (err, ctx) => {
  if (ctx.ext.failToast) {
    const { message, retcode, type, statusCode } = err || {};
    const codeText = retcode ? `(${retcode})` : '';

    if (message) {
      showToast({
        title: type === RequestError.fail ? '请求失败' : `${message}${codeText}`,
        icon: 'none',
      });
    }
  }
};

export function gGet<T>(data: IRequestOptions<Partial<IExt>>) {
  return gRequest.get<IRequestRes<T>>(data);
}

export function gPost<T>(data: IRequestOptions<Partial<IExt>>) {
  return gRequest.post<IRequestRes<T>>(data);
}

export default gRequest;
```

## 注意事项

- 该库 TS 编译的 `target` 为 `ES2015`。如果要兼容到老版本，请再进行一次 babel 编译。


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## 2.0.0 (2022-10-24)

### Features

- 导出更多函数, 方便扩展 Request 类 ([6376835](https://github.com/marvin1023/cloud-sea/commit/63768359357604ea2dc7fd9b1e5e5fbf12f9dad2))
- 默认配置参数通过函数获取, 方便于扩展类 ([9a1d618](https://github.com/marvin1023/cloud-sea/commit/9a1d618bc8e3e4e75f583936c23438c4cac54a1f))
- 重新梳理逻辑及 TS ([af525b2](https://github.com/marvin1023/cloud-sea/commit/af525b2d58f2fab0a0ca622c0a6a60cdf618cbfb))
- ext 新增 urlHasNoSearch 属性,表示不带 query 的 url ([d3a5a97](https://github.com/marvin1023/cloud-sea/commit/d3a5a97dee2bb61d1d033caa8dd44b3d53d2fcc3))
- init ([563a1a2](https://github.com/marvin1023/cloud-sea/commit/563a1a24e2d770af5510bb20d63d89a41f3d5a54))

### Bug Fixes

- 调整 script 及删除垃圾信息 ([8d2806a](https://github.com/marvin1023/cloud-sea/commit/8d2806a10e61349ed143d766c7127c63778574c4))
- 默认 taskName 使用自动 id ([2a42482](https://github.com/marvin1023/cloud-sea/commit/2a42482068838bd88a4848f44c39e398a0702f9b))
- 修复 main 入口等 ([da9c314](https://github.com/marvin1023/cloud-sea/commit/da9c314d38342ecfd4b612863c88c4a60c121c20))


src/types.ts
export type IAnyObject = Record<string, any>;
export type IMethod = 'OPTIONS' | 'GET' | 'HEAD' | 'POST' | 'PUT' | 'DELETE' | 'TRACE' | 'CONNECT';

export interface IWXReqOptions {
  url: string;
  data?: string | IAnyObject | ArrayBuffer;
  header?: IAnyObject;
  timeout?: number;
  method?: IMethod;
  dataType?: 'json' | '其他';
  enableCache?: boolean;
  enableChunked?: boolean;
  enableHttp2?: boolean;
  enableHttpDNS?: boolean;
  enableQuic?: boolean;
  forceCellularNetwork?: boolean;
  httpDNSServiceId?: string;
  responseType?: 'text' | 'arraybuffer';
}

export interface IXHRReqOptions {
  url: string;
  method?: IMethod;
  header?: Record<string, string>;
  data?: Document | XMLHttpRequestBodyInit | null | undefined | IAnyObject;
  timeout?: number;
  withCredentials?: boolean;
  responseType?: 'text' | 'arraybuffer' | 'blob' | 'json' | '' | 'document';
  async?: boolean;
}

export interface IWXRequestSuccessCallbackResult {
  cookies?: string[];
  data: IAnyObject;
  header: IAnyObject;
  statusCode: number;
  errMsg: string;
}

export interface IXHRRequestSuccessCallbackResult {
  statusCode: number;
  data: IAnyObject;
  header: IAnyObject;
}

export interface IWXRequestFailCallbackResult {
  errMsg: string;
}

export type IAdapter = (
  config: IAnyObject & { url: string },
  resolve: (value: unknown) => void,
  reject: (reason?: any) => void,
) => any;

export type IReqOptions = IWXReqOptions | IXHRReqOptions;
export type IRequestSuccessCallbackResult = IWXRequestSuccessCallbackResult | IXHRRequestSuccessCallbackResult;

export interface ICloudSeaConfig {
  baseUrl: string; // 基础 url，以 https 开头
  repeatNum: number; // 请求失败重试次数
  xRequestId: boolean; // 是否生成请求 id
  timeout: number; // 超时时间，如果确定发请求的 api 本身就支持 timeout 属性，可以设置该值为 0
  retcodeKey: false | string | string[]; // retcode 字段，false 表示不启用该功能
  retcodeWhiteList: false | number[]; // retcode 白名单，默认 0 和 白名单表示业务成功，其余为失败，false 表示不启用该功能。
  logicErrorMsgKey: string; // 业务逻辑错误文本字段
  logicErrorMsgUnknown: string; // 默认的业务逻辑错误文本，如果后台没有返回对应的错误信息，则将使用该信息
  adapter: IAdapter | undefined; // 请求 adapter
}

export interface IRequestTask {
  name?: string; // 用于取消请求
  client: any;
  urlHasNoSearch: string; // 去掉请求 url 的 query，可用于上报请求地址
  timer: ReturnType<typeof setTimeout>;
  costTime: number; // 请求总共花费时间，当 xRequestTime 为 true，则有该值
  repeatTry: () => Promise<IAnyObject>; // 用于重试
}

export interface IRequestCtx<T extends ICloudSeaConfig> {
  req: IReqOptions & { header: Record<string, string> };
  res: IRequestSuccessCallbackResult | Record<string, never>;
  ext: T;
  task: IRequestTask;
}

export type IRequestOptions<T extends Partial<ICloudSeaConfig>> = IReqOptions & {
  ext?: T;
} & { task?: { name: string } };

export type ICloudSeaOptions<T extends Partial<ICloudSeaConfig>> = Omit<IRequestOptions<T>, 'url' | 'data' | 'task'>;

export type IRequestPluginFn<U extends ICloudSeaConfig> = (ctx: IRequestCtx<U>) => IRequestCtx<U>;


src/core/cloud-sea.ts
import { IRequestCtx, ICloudSeaOptions, IAnyObject, IRequestOptions, IMethod, ICloudSeaConfig } from '../types';
import { RequestError } from './requestError';
import { defaultConfig } from './default';
import { Plugins } from './plugins';
import { Task } from './task';

export class CloudSea<U extends ICloudSeaConfig = ICloudSeaConfig> {
  public static readonly config: ICloudSeaConfig = { ...defaultConfig };

  public setConfig = (config: Partial<ICloudSeaConfig>) => {
    Object.assign(CloudSea.config, config);
  };

  private options: ICloudSeaOptions<Partial<U>>;
  public res: Plugins<IRequestCtx<U>>;
  public req: Plugins<IRequestCtx<U>>;
  public taskRequesting: Record<string, Task<IRequestCtx<U>>> = {};

  public completeHandler?(ctx: IRequestCtx<U>, err?: RequestError): void;
  public thenHandler?(ctx: IRequestCtx<U>): IAnyObject;
  public catchHandler?(err: RequestError, ctx: IRequestCtx<U>): void | RequestError;

  public constructor(options?: ICloudSeaOptions<Partial<U>>) {
    this.initOptions(options);
    this.initPlugins();
  }

  public initOptions(options?: ICloudSeaOptions<Partial<U>>) {
    this.options = options || {};
  }

  private initPlugins() {
    this.req = new Plugins<IRequestCtx<U>>();
    this.res = new Plugins<IRequestCtx<U>>();
  }

  public request<T>(data: IRequestOptions<Partial<U>>): Promise<T> {
    const { config } = CloudSea;
    const { ext: optionsExt, ...optionsReq } = this.options;
    const { ext: dataExt, task: dataTask, ...dataReq } = data;

    const ext = { ...config, ...optionsExt, ...dataExt };

    const header = this.reqHeaderHandler(ext.xRequestId);

    const newData = {
      req: { header, ...optionsReq, ...dataReq },
      ext,
      task: dataTask || {},
    } as Omit<IRequestCtx<U>, 'res'>;

    const requestTask = new Task<IRequestCtx<U>>(newData);

    if (dataTask?.name) {
      this.taskRequesting[dataTask.name] = requestTask;
    }

    this.res.pipe(requestTask.ctx);

    return requestTask
      .dispatch()
      .then((res) => {
        this.taskComplete(requestTask.ctx);
        this.res.pipe(requestTask.ctx);
        return (this.thenHandler?.(requestTask.ctx) || res) as T;
      })
      .catch((err: RequestError) => {
        this.taskComplete(requestTask.ctx, err);
        const error = this.catchHandler?.(err, requestTask.ctx);
        return Promise.reject(error || err);
      });
  }

  public method<T>(method: IMethod, data: IRequestOptions<Partial<U>>) {
    return this.request<T>({ ...data, method });
  }

  public get<T>(data: IRequestOptions<Partial<U>>) {
    return this.method<T>('GET', data);
  }

  public post<T>(data: IRequestOptions<Partial<U>>) {
    return this.method<T>('POST', data);
  }

  public abort(taskName?: string) {
    // abort 单个
    if (taskName) {
      this.taskRequesting[taskName]?.abort();
      return;
    }

    // abort 所有
    Object.keys(this.taskRequesting).forEach((item) => {
      this.taskRequesting[item].abort();
    });
  }

  protected taskComplete(ctx: IRequestCtx<U>, err?: RequestError) {
    const { name } = ctx.task;
    if (name && this.taskRequesting[name]) {
      delete this.taskRequesting[name];
    }

    this.completeHandler?.(ctx, err);
  }

  protected reqHeaderHandler(xRequestId: boolean) {
    const header = {};

    // request-id
    if (xRequestId) {
      Object.assign(header, { 'X-Request-Id': this.getUUID() });
    }

    // 自定义请求发起时间
    Object.assign(header, { 'X-Request-Time': Date.now() });

    return header;
  }

  protected getUUID(bytes = 16) {
    const SHARED_CHAR_CODES_ARRAY = Array(32);
    for (let i = 0; i < bytes * 2; i++) {
      SHARED_CHAR_CODES_ARRAY[i] = Math.floor(Math.random() * 16) + 48;
      // valid hex characters in the range 48-57 and 97-102
      if (SHARED_CHAR_CODES_ARRAY[i] >= 58) {
        SHARED_CHAR_CODES_ARRAY[i] += 39;
      }
    }
    return String.fromCharCode.apply(null, SHARED_CHAR_CODES_ARRAY.slice(0, bytes * 2));
  }
}


src/core/task.ts
import { IRequestCtx, IAnyObject, ICloudSeaConfig } from '../types';
import { RequestError } from './requestError';
import { REQUEST_ERROR_MAP, DEFAULT_RETCODE_KEY } from './const';

export class Task<X extends IRequestCtx<ICloudSeaConfig>> {
  public ctx: X;

  public constructor(options: Omit<X, 'res'>) {
    this.init(options);
  }

  private init(options: Omit<X, 'res'>) {
    this.initCtx(options);
  }

  private initCtx(opts: Omit<X, 'res'>) {
    const { ext, req, task } = opts;

    // 构造 ctx 对象
    this.ctx = {
      req,
      res: {},
      ext,
      task,
    } as unknown as X;
  }

  public dispatch(): Promise<X> {
    const { ctx } = this;
    const { timeout } = ctx.ext;

    // 处理 url
    this.urlHandler();

    // 重试逻辑处理
    const repeatTry = (): Promise<X> => {
      // 超时 abort 处理
      if (timeout) {
        ctx.task.timer = setTimeout(() => {
          ctx.task.client?.abort();
          clearTimeout(ctx.task.timer);
        }, timeout);
      }

      return this.promiseAdapter()
        .then((result) => {
          const { statusCode } = result;
          if (statusCode >= 200 && statusCode < 300) {
            this.clearHandler();
            // 挂到 ctx.res 上
            this.ctx.res = result;

            // retcode 处理
            this.retcodeHandler();
            this.retcodeWhiteListHandler();

            return this.ctx;
          }

          // 处理服务器错误 message
          throw new RequestError('Server Error', { type: REQUEST_ERROR_MAP.server, statusCode });
        })
        .catch((err) => {
          // 非逻辑错误，执行清空操作及重试
          if (!(err.type && err.type === REQUEST_ERROR_MAP.logic)) {
            this.clearHandler();

            // 重试
            if (this.ctx.ext.repeatNum && this.ctx.ext.repeatNum--) {
              return repeatTry();
            }
          }

          const newError = err.type
            ? err
            : new RequestError(err.message || err.errMsg, { type: REQUEST_ERROR_MAP.fail });

          throw newError;
        });
    };

    this.ctx.task.repeatTry = repeatTry;

    return repeatTry();
  }

  public abort() {
    this.ctx.task.client?.abort();
  }

  protected promiseAdapter(): Promise<any> {
    const { req, ext } = this.ctx;
    const { adapter } = ext;

    if (!adapter) {
      return Promise.reject('adapter is must be required');
    }

    return new Promise((resolve, reject) => {
      this.ctx.task.client = adapter(req, resolve, reject);
    });
  }

  protected clearHandler() {
    const { ctx } = this;
    const { timeout } = ctx.ext;
    const { timer } = ctx.task;

    // 计算请求耗时
    ctx.task.costTime = Date.now() - ctx.req.header['X-Request-Time'];

    // 清除计时
    if (timeout && timer) {
      clearTimeout(timer);
    }
  }

  protected retcodeHandler() {
    const { ctx } = this;
    const { retcodeKey } = ctx.ext;
    if (!retcodeKey) {
      return;
    }
    // 如果key值为非 retcode 的 string，则手动添加 retcode 等于该 key 的值
    if (typeof retcodeKey === 'string' && retcodeKey !== DEFAULT_RETCODE_KEY) {
      ctx.res.data[DEFAULT_RETCODE_KEY] = ctx.res.data[retcodeKey];
      return;
    }

    // 如果 key 值为数组，如['retcode', 'code', 'ret_code', 'ret'];
    if (Array.isArray(retcodeKey)) {
      const len = retcodeKey.length;

      for (let i = 0; i < len; i++) {
        const item = ctx.res.data[retcodeKey[i]];
        if (item !== undefined) {
          ctx.res.data[DEFAULT_RETCODE_KEY] = item;
          break;
        }
      }
    }
  }

  protected retcodeWhiteListHandler() {
    const { ctx } = this;
    const {
      ext: { retcodeKey, retcodeWhiteList },
      res,
    } = ctx;

    // 关闭白名单，retcode 不论为啥都为成功
    // 或者没有 retcodeKey
    if (!retcodeWhiteList || !retcodeKey) {
      return;
    }

    // 如果为 0 或者在白名单内，则进入 then 处理
    const retArr = retcodeWhiteList;
    const { retcode } = res.data || {};
    const isWhite = retArr.includes(retcode);
    if (retcode === 0 || isWhite) {
      return;
    }

    // 逻辑错误
    const logicErrMsg = this.getLogicErrMsg();
    throw new RequestError(logicErrMsg, {
      type: REQUEST_ERROR_MAP.logic,
      retcode,
    });
  }

  protected getLogicErrMsg(): string {
    const { ctx } = this;
    const { res, ext } = ctx;
    const { logicErrorMsgKey, logicErrorMsgUnknown } = ext;

    // 如果没有指定 msg 的 key
    if (!logicErrorMsgKey) {
      return logicErrorMsgUnknown;
    }

    // 如果有指定，则取该值
    let logicErrMsg: string | undefined = res.data?.[logicErrorMsgKey];
    // 错误信息可能不是一个直接的 string 字段，而是被包裹在一个对象中，如 errData: { text: 'xxx', code: xxx };
    // 这样可以设置 key 值为 'errData.text'，分割得到最终的字段
    if (!logicErrMsg && logicErrorMsgKey.includes('.')) {
      const [key1, key2] = logicErrorMsgKey.split('.');
      if (res.data?.[key1]?.[key2]) {
        logicErrMsg = res.data[key1][key2];
      }
    }

    return logicErrMsg || logicErrorMsgUnknown;
  }

  protected urlHandler() {
    const { ctx } = this;
    const {
      ext: { baseUrl },
      req: { url },
    } = ctx;
    if (baseUrl && !url.startsWith('http')) {
      ctx.req.url = `${baseUrl}${url}`;
    }

    ctx.task.urlHasNoSearch = ctx.req.url.split('?')[0]; // 不带 query 的 url，可用于统计或上报等
  }
}
